lib/ansible/galaxy/collection/__init__.py

```diff
@@ -559,34 +218,45 @@ def build_collection(collection_path, output_path, force):
     :return: The path to the collection build artifact.
     """
     b_collection_path = to_bytes(collection_path, errors='surrogate_or_strict')
-    b_galaxy_path = get_galaxy_metadata_path(b_collection_path)
-    if not os.path.exists(b_galaxy_path):
-        raise AnsibleError("The collection galaxy.yml path '%s' does not exist." % to_native(b_galaxy_path))
-
-    info = CollectionRequirement.galaxy_metadata(b_collection_path)
-
-    collection_manifest = info['manifest_file']
-    collection_meta = collection_manifest['collection_info']
-    file_manifest = info['files_file']
+    collection_meta = _get_meta_from_src_dir(b_collection_path)
+    collection_manifest = _build_manifest(**collection_meta)
+    file_manifest = _build_files_manifest(
+        b_collection_path,
+        collection_meta['namespace'],  # type: ignore
+        collection_meta['name'],  # type: ignore
+        collection_meta['build_ignore'],  # type: ignore
+    )
 
-    collection_output = os.path.join(output_path, "%s-%s-%s.tar.gz" % (collection_meta['namespace'],
-                                                                       collection_meta['name'],
-                                                                       collection_meta['version']))
+    artifact_tarball_file_name = '{ns!s}-{name!s}-{ver!s}.tar.gz'.format(
+        name=collection_meta['name'],
+        ns=collection_meta['namespace'],
+        ver=collection_meta['version'],
+    )
+    b_collection_output = os.path.join(
+        output_path,
```

Comment:
This needs to be a byte string so you aren't trying to join a str and bytes together.

```suggestion
        to_bytes(output_path, errors='surrogate_or_strict'),
```

Reply:
fixed by passing bytes to this function